File: ios/Classes/FloatingLabel.swift
//
//  FloatingLabel.swift
//  mapbox_gl
//
//  Created by Eugene Pisotsky on 20.10.2019.
//

import Foundation
import Mapbox

class FloatingLabel {
    var ID: String
    var SOURCE_ID: String
    var LAYER_ID: String
    var IMAGE_ID: String
    
    var mapInstance: MGLMapView
    var id: String
    var image: UIImage
    var location: CLLocationCoordinate2D
    var width: Float
    var height: Float
    var icon: String?
    var style: MGLStyle?
   
    var anchorX: Float = 1
    var anchorY: Float = 1
    
    var translateY: Double
    var translateX: Double
    
    var pointSource: MGLShapeSource
    var labelSource: MGLShapeSource
    
    var pointLayer: MGLSymbolStyleLayer
    var labelLayer: MGLSymbolStyleLayer
    
    init(
        mapInstance: MGLMapView,
        id: String,
        image: UIImage,
        location: CLLocationCoordinate2D,
        width: Float,
        height: Float,
        icon: String?
    ) {
        self.mapInstance = mapInstance
        self.id = id
        self.image = image
        self.location = location
        self.width = width
        self.height = height
        self.icon = icon
        
        style = mapInstance.style
        
        ID = id
        SOURCE_ID = id + "-source"
        LAYER_ID = id + "-layer"
        IMAGE_ID = id + "-image"
        
        let point = MGLPointFeature()
        point.identifier = id
        point.coordinate = location
        
        pointSource = MGLShapeSource(identifier: "point-" + SOURCE_ID, shape: point, options: nil)
        labelSource = MGLShapeSource(identifier: SOURCE_ID, shape: point, options: nil)
        pointLayer = MGLSymbolStyleLayer(identifier: "point-" + LAYER_ID, source: pointSource)
        labelLayer = MGLSymbolStyleLayer(identifier: LAYER_ID, source: labelSource)
        
        style?.addSource(pointSource)
        style?.addSource(labelSource)
        style?.setImage(image, forName: IMAGE_ID)
        
        translateX = Double((width - 36) / 2);
        translateY = Double((height / 2) + 4);
        
        pointLayer.iconAnchor = NSExpression(forConstantValue: "center")
        pointLayer.iconImageName = NSExpression(forConstantValue: icon)
        pointLayer.iconAllowsOverlap = NSExpression(forConstantValue: true)
        
        labelLayer.iconImageName = NSExpression(forConstantValue: IMAGE_ID)
        labelLayer.iconTranslation = NSExpression(forConstantValue: NSValue(cgVector: CGVector(dx: translateX, dy: translateY)))
        labelLayer.iconAllowsOverlap = NSExpression(forConstantValue: true)
        
        style?.addLayer(pointLayer)
        style?.addLayer(labelLayer)
    }
    
    func getId() -> String {
        return self.id
    }
    
    func getLayerId() -> String {
        return LAYER_ID
    }
    
    func destroy() {
        style?.removeLayer(pointLayer)
        style?.removeLayer(labelLayer)
        style?.removeImage(forName: IMAGE_ID)
        style?.removeSource(pointSource)
        style?.removeSource(labelSource)
    }
    
    func hide() {
        pointLayer.isVisible = false
        labelLayer.isVisible = false
    }
    
    func show() {
        pointLayer.isVisible = true
        labelLayer.isVisible = true
    }
    
    func updatePosition() {
        let screenSize: CGRect = UIScreen.main.bounds
        let windowWidth = screenSize.size.width
        let windowHeight = screenSize.size.height - CGFloat(mapInstance.contentInset.bottom)

        let centerScreenPoint: CGPoint = self.mapInstance.convert(location, toPointTo: nil)

        let x = (centerScreenPoint.x + CGFloat(width))
        let y = (centerScreenPoint.y - CGFloat(height))
        
        let x2 = (centerScreenPoint.x - CGFloat(width))
        let y2 = (centerScreenPoint.y + CGFloat(height)) - CGFloat(mapInstance.contentInset.top)

        if (x > windowWidth) {
            translateX = Double(-((width - 36) / 2))
            anchorX = -1
        } else if (x2 < 0) {
            translateX = Double((width - 36) / 2)
            anchorX = 1
        }
        
        if (y > windowHeight / 2) {
            translateY = Double(-((height / 2) - 12))
            anchorY = -1
        } else if (y2 < windowHeight / 2) {
            translateY = Double((height / 2) + 4)
            anchorY = 1
        }
        
        labelLayer.iconTranslation = NSExpression(forConstantValue: NSValue(cgVector: CGVector(dx: translateX, dy: translateY)))
    }
    
    func updateSourceCoordinates(coords: CLLocationCoordinate2D) {
        self.location = coords
        let point = MGLPointFeature()
        point.identifier = self.ID
        point.coordinate = location
        
        pointSource.shape = point
        labelSource.shape = point
    }
    
    func updateIcon(icon: String) {
        pointLayer.iconImageName = NSExpression(forConstantValue: icon)
    }
    
    func updateLabel(
        width: Float,
        height: Float,
        image: UIImage
    ) {
        self.width = width
        self.height = height
        
        style?.removeImage(forName: IMAGE_ID)
        style?.setImage(image, forName: IMAGE_ID)
        
        updatePosition()
    }
    
    func getAnchor() -> [Float] {
        return [anchorX, anchorY]
    }
    
}

File: android/src/main/java/com/mapbox/mapboxgl/FloatingLabel.java
package com.mapbox.mapboxgl;

import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.PointF;
import android.view.Display;

import com.mapbox.geojson.Feature;
import com.mapbox.geojson.Point;
import com.mapbox.mapboxsdk.geometry.LatLng;
import com.mapbox.mapboxsdk.maps.MapView;
import com.mapbox.mapboxsdk.maps.MapboxMap;
import com.mapbox.mapboxsdk.maps.Style;
import com.mapbox.mapboxsdk.style.layers.Property;
import com.mapbox.mapboxsdk.style.layers.PropertyFactory;
import com.mapbox.mapboxsdk.style.layers.SymbolLayer;
import com.mapbox.mapboxsdk.style.sources.GeoJsonSource;

import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.iconAllowOverlap;
import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.iconAnchor;
import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.iconSize;
import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.iconTranslate;
import static com.mapbox.mapboxsdk.style.layers.PropertyFactory.visibility;

public class FloatingLabel {

    private Style style;

    private float anchorX = 1f;
    private float anchorY = 1f;

    private float translateX;
    private float translateY;

    private float prevTranslateX;
    private float prevTranslateY;

    private String ID;
    private String SOURCE_ID;
    private String LAYER_ID;
    private String IMAGE_ID;

    private LatLng location;
    private Activity activity;
    private MapView mapboxMapView;
    private MapboxMap mapboxMap;
    private float density;
    private float width;
    private float height;
    private String icon;

    private GeoJsonSource pointSource;
    private GeoJsonSource labelSource;

    private LatLng prevCameraPosition;

    private MapConfiguration mapConfiguration;

    public FloatingLabel(
            MapView mapView,
            MapboxMap mapboxMap,
            String id,
            Bitmap image,
            LatLng location,
            float width,
            float height,
            String icon,
            float density,
            Activity activity,
            MapConfiguration mapConfiguration
    ) {
        this.mapboxMapView = mapView;
        this.width = width;
        this.height = height;
        this.location = location;
        this.activity = activity;
        this.mapboxMap = mapboxMap;
        this.density = density;
        this.icon = icon;
        this.mapConfiguration = mapConfiguration;

        ID = id;
        SOURCE_ID = id + "-source";
        LAYER_ID = id + "-layer";
        IMAGE_ID = id + "-image";

        style = mapboxMap.getStyle();

        pointSource = new GeoJsonSource(
                "point-" + SOURCE_ID,
                Feature.fromGeometry(Point.fromLngLat(location.getLongitude(), location.getLatitude()))
        );

        labelSource = new GeoJsonSource(
                SOURCE_ID,
                Feature.fromGeometry(Point.fromLngLat(location.getLongitude(), location.getLatitude()))
        );

        style.addSource(pointSource);
        style.addSource(labelSource);
        style.addImage(IMAGE_ID, image);

        translateX = (width - 36) / 2;
        translateY = (height / 2) + 4;

        style.addLayer(new SymbolLayer(LAYER_ID, SOURCE_ID)
                .withProperties(
                        PropertyFactory.iconImage(IMAGE_ID),
                        iconAllowOverlap(true),
                        iconAnchor(Property.ICON_ANCHOR_CENTER),
                        iconTranslate(new Float[]{translateX, translateY})
                )
        );

        style.addLayer(new SymbolLayer("point-" + LAYER_ID, "point-" + SOURCE_ID)
                .withProperties(
                        PropertyFactory.iconImage(icon),
                        iconAllowOverlap(true),
                        iconSize(1.2f),
                        iconAnchor(Property.ICON_ANCHOR_CENTER)
                )
        );
    }

    public void updatePosition() {
        activity.runOnUiThread(new Runnable() {
            public void run() {
                final LatLng target = mapboxMap.getCameraPosition().target;

                if (style.isFullyLoaded() && !target.equals(prevCameraPosition)) {
                    prevCameraPosition = target;

                    Display display = activity.getWindowManager().getDefaultDisplay();
                    android.graphics.Point size = new android.graphics.Point();
                    display.getSize(size);
                    int windowWidth = size.x;
                    int windowHeight = size.y - (int) mapConfiguration.bottomPadding;

                    final PointF pointInWindow = mapboxMap.getProjection().toScreenLocation(location);
                    final float x = (pointInWindow.x + (width * density));
                    final float y = (pointInWindow.y - (height * density));

                    final float x2 = (pointInWindow.x - (width * density));
                    final float y2 = (pointInWindow.y + (height * density)) - (int) mapConfiguration.topPadding;

                    if (x > windowWidth) {
                        translateX = -((width - 36) / 2);
                        anchorX = -1;
                    } else if (x2 < 0) {
                        translateX = (width - 36) / 2;
                        anchorX = 1;
                    }

                    if (y > windowHeight / 2) {
                        translateY = -((height / 2) - 12);
                        anchorY = -1;
                    } else if (y2 < windowHeight / 2) {
                        translateY = (height / 2) + 4;
                        anchorY = 1;
                    }

                    if (translateX != prevTranslateX || translateY != prevTranslateY) {
                        prevTranslateX = translateX;
                        prevTranslateY = translateY;

                        style.getLayerAs(LAYER_ID).setProperties(
                                iconTranslate(new Float[]{translateX, translateY})
                        );
                    }
                }
            }
        });
    }

    public void destroy() {
        style.removeLayer(LAYER_ID);
        style.removeLayer("point-" + LAYER_ID);
        style.removeImage(IMAGE_ID);
        style.removeSource(SOURCE_ID);
        style.removeSource("point-" + SOURCE_ID);
    }

    public String getLayerId() {
        return LAYER_ID;
    }

    public String getId() {
        return ID;
    }

    public void updateSourceCoordinates(LatLng location) {
        this.location = location;

        final Feature point = Feature.fromGeometry(Point.fromLngLat(location.getLongitude(), location.getLatitude()));

        pointSource.setGeoJson(point);
        labelSource.setGeoJson(point);
    }

    public void updateLabel(float width, float height, Bitmap image) {
        this.width = width;
        this.height = height;

        style.removeImage(IMAGE_ID);
        style.addImage(IMAGE_ID, image);

        updatePosition();
    }

    public void updateIcon(String icon) {
        // @todo
    }

    public void hide() {
        style.getLayer(LAYER_ID).setProperties(visibility(Property.NONE));
        style.getLayer("point-" + LAYER_ID).setProperties(visibility(Property.NONE));
    }

    public void show() {
        style.getLayer(LAYER_ID).setProperties(visibility(Property.VISIBLE));
        style.getLayer("point-" + LAYER_ID).setProperties(visibility(Property.VISIBLE));
    }

    public PointF getAnchor() {
        return new PointF(anchorX, anchorY);
    }

}